File: leetcode/102.binary-tree-level-order-traversal.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: TreeNode | None) -> list[list[int]]:
        level = [root]
        res: list[list[int]] = []

        while level:
            level_values: list[int] = []
            next_level: list[TreeNode | None] = []

            for node in level:
                if not node:
                    continue

                level_values.append(node.val)
                next_level.extend([node.left, node.right])

            if level_values:
                res.append(level_values)

            level = next_level

        return res


# @leet end

File: leetcode/102.binary-tree-level-order-traversal.cpp
// @leet start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <queue>
#include <vector>

class Solution {
 public:
  std::vector<std::vector<int>> levelOrder(TreeNode* root) {
    std::vector<std::vector<int>> result;
    std::queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
      const int levelSize = q.size();
      std::vector<int> level;

      for (int _ = 0; _ < levelSize; _++) {
        const auto node = q.front();
        q.pop();

        if (!node) {
          continue;
        }

        level.push_back(node->val);

        q.push(node->left);
        q.push(node->right);
      }

      if (!level.empty()) {
        result.push_back(level);
      }
    }

    return result;
  }
};
// @leet end
